File: Medium/0133-clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/
class Solution {
public:
    Node* cloneGraph(Node* node) {
        unordered_map<int, Node*> mp;
        return node == nullptr ? nullptr : helper(mp, node);
    }

    Node* helper(unordered_map<int, Node*>& mp, Node* node) {
        Node* p = new Node(node->val);
        mp.emplace(node->val, p);
        unordered_map<int, Node*>::iterator it;
        for (int i = 0; i < node->neighbors.size(); i++) {
            it = mp.find(node->neighbors[i]->val);
            if (it == mp.end()) {
                p->neighbors.push_back(helper(mp, node->neighbors[i]));
            } else {
                p->neighbors.push_back(it->second);
            }
        }
        return p;
    }
};

File: Medium/0133-clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    private Map<Integer, Node> seen;

    public Node doClone(Node node) {
        if (node == null) {
            return null;
        }
        if (seen.containsKey(node.val)) {
            return seen.get(node.val);
        }
        ArrayList<Node> neighbors = new ArrayList<>();
        Node newNode = new Node(node.val, neighbors);
        seen.put(node.val, newNode);
        for (Node neighbor : node.neighbors) {
            neighbors.add(doClone(neighbor));
        }
        return newNode;
    }

    public Node cloneGraph(Node node) {
        seen = new HashMap<>();
        return doClone(node);
    }
}
